src/passes/MinimizeRecGroups.cpp

```diff
@@ -0,0 +1,752 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Split types into minimal recursion groups by computing the strongly connected
+// components of the type graph, which are the minimal sets of
+// mutually-recursive types that must be in the same recursion group with each
+// other for the type section to validate.
+//
+// We must additionally ensure that distinct types remain distinct, which would
+// not be the case if we allowed two separate strongly connected components with
+// the same shape to be rewritten to two copies of the same recursion group.
+// Accidentally merging types would be incorrect because it would change the
+// behavior of casts that would have previously been able to differentiate the
+// types.
+//
+// When multiple strongly connected components have the same shape, first try to
+// differentiate them by permuting their types, which keeps the sizes of the rec
+// groups as small as possible. When there are more strongly connected
+// components that are permutations of one another than there are distinct
+// permutations, fall back to keeping the types distinct by adding distinct
+// brand types to the recursion groups to ensure they have different shapes.
+//
+// There are several possible algorithmic design for detecting when to generate
+// permutations and determining what permutations to generate. They trade off
+// the amount of "wasted" work spent generating permutations that have already
+// been used with the amount of work spent trying to avoid the wasted work and
+// with the quality of the output.
+//
+// The simplest possible design that avoids all "wasted" work would be to
+// canonicalize the type of each group to eagerly detect isomorphisms and
+// eagerly organize the groups into equivalence classes. This would allow us to
+// avoid ever generating permutations of a group that have already been used.
+// See `getCanonicalPermutation` below for details on how this works. However,
+// this is not an ideal solution because canonicalization is an expensive
+// operation and in the common case a group will not conflict with any other
+// group, so canonicalization itself would be wasted work.
+//
+// The simplest possible design in the other direction is to never canonicalize
+// and instead to lazily build up equivalences classes of isormorphic groups
+// when conflicts are detected in practice. Conflicts are resolved by choosing
+// the next permutation generated by the representative element of the
+// equivalence class. This solution is not ideal because groups with high
```

Comment:
I see, thanks. In that case, is it worth mentioning it here? That we store that on the representative element is an implementation detail, not fundamental to the algorithm. I feel that keeping this comment high level would be clearer, in other words.

Reply:
I think this detail is useful because it clarifies what happens when equivalence classes are merged. After a merge, the subsequent permutations are generated by whatever the new representative element is. If we take these comments as roadmaps for how we might rewrite and simplify the pass in the future, that level of detail could be useful.